.github/workflows/test-deploy.yml

```diff
@@ -0,0 +1,139 @@
+          registry-url: https://registry.npmjs.org
+      - run: npm install
+      - run:
+          npm ${{ github.event_name == 'release' && 'publish' || 'publish
+          --dry-run' }}
+        env:
+          NODE_AUTH_TOKEN: ${{ secrets.NPM_TOKEN }}
+  images:
+    runs-on: ubuntu-latest
+    needs: packages
+    strategy:
+      matrix:
+        cml: [0]
+        dvc: [1, 2]
+        base: [0, 1]
+        gpu: [false, true]
+        include:
+          - base: 0
+            ubuntu: 18.04
+            python: 2.7
+            cuda: 10.1
+            cudnn: 7
+          - base: 1
+            ubuntu: 20.04
+            python: 3.8
+            cuda: 11.0.3
+            cudnn: 8
+    steps:
+      - uses: actions/checkout@v2
+      - uses: docker/setup-buildx-action@v1
+      - uses: actions/cache@v2
+        with:
+          path: /tmp/.buildx-cache
+          key: ${{ runner.os }}-buildx-${{ github.sha }}
+          restore-keys: ${{ runner.os }}-buildx-
+      - uses: docker/login-action@v1
+        with:
+          registry: docker.io
+          username: ${{ secrets.DOCKERHUB_USERNAME }}
+          password: ${{ secrets.DOCKERHUB_PASSWORD }}
+      - uses: docker/login-action@v1
+        with:
+          registry: ghcr.io
+          username: ${{ github.repository_owner }}
+          password: ${{ github.token }}
+      - uses: docker/build-push-action@v2
+        if: matrix.gpu == false
+        with:
+          push:
+            ${{ github.event_name == 'release' || github.event_name ==
+            'schedule' }}
+          context: ./
+          file: ./Dockerfile
+          tags: |
+            docker.io/dvcorg/cml:${{ matrix.cml }}-dvc${{ matrix.dvc }}-base${{
+            matrix.base }}
+            ghcr.io/iterative/cml:${{ matrix.cml }}-dvc${{ matrix.dvc }}-base${{
+            matrix.base }}
+          build-args: |
+            CML_VERSION=${{ matrix.cml }}
+            DVC_VERSION=${{ matrix.dvc }}
+            PYTHON_VERSION=${{ matrix.python }}
+            BASE_IMAGE=ubuntu:${{ matrix.ubuntu }}
+          pull: true
+          cache-from: type=local,src=/tmp/.buildx-cache
+          cache-to: type=local,dest=/tmp/.buildx-cache-new
+      - uses: docker/build-push-action@v2
+        if: matrix.gpu == true
+        with:
+          push:
+            ${{ github.event_name == 'release' || github.event_name ==
+            'schedule' }}
+          context: ./
+          file: ./Dockerfile
+          tags: |
+            docker.io/dvcorg/cml:${{ matrix.cml }}-dvc${{ matrix.dvc }}-base${{
+            matrix.base }}-gpu
+            ghcr.io/iterative/cml:${{ matrix.cml }}-dvc${{ matrix.dvc }}-base${{
+            matrix.base }}-gpu
+          build-args: |
+            CML_VERSION=${{ matrix.cml }}
+            DVC_VERSION=${{ matrix.dvc }}
+            PYTHON_VERSION=${{ matrix.python }}
+            BASE_IMAGE=nvidia/cuda:${{ matrix.cuda }}-cudnn${{ matrix.cudnn }}-runtime-ubuntu${{ matrix.ubuntu }}
+          pull: true
```

Comment:
`pull: true` should fetch any updates to the base images, potentially invalidating the cache as necessary.

Reply:
regarding the CML version... Hmm, yes, this is a problem.

`cml: [0]` is problematic.

We really should be building `cml: [<current_commit>]`, no?

Reply:
> _We really should be building `cml: [<current_commit>]` no?_

Yes, but indirectly: the `images` job `needs` the `packages` job so we pull the package after it has been published; hacky, but compatible with major version bumps.

Reply:
how about now? :)

Reply:
I find great the deduplication, but still struggle to understand the implications of this new approach:

- The `RUN` command that installs DVC is always the same for a given major version, and caching that layer would effectively prevent it from updating each time a new minor or patch version gets released. _Id est: the `RUN` command produces a layer reference derived from the command itself and not by the output it produces, so `RUN install dvc@1` will hit the cache no matter if it resolves to DVC 1.0 or 1.1 through the package manager._ 

- Invalidating the cache each time we release CML implies that DVC could not be updated independently with a `cron` job, as the version of the former would always be the same and point to the same cached artifacts, causing the latter to be pulled from the cache.

Reply:
> caching that layer would effectively prevent it from updating each time a new minor or patch version gets released

the cache for that layer would not be available (the cache is keyed on the the full version)

> DVC could not be updated independently with a `cron` job

never realised this was one of the intentions. The way it is atm, the `cron` job will update everything when the base images are updated (e.g. `ubuntu` updates its `20.04` tag - which is very likely).

If we care about DVC version in the cron job, we'll have to add its full version to the cache key too.

Reply:
> _the cache is keyed on the the full version_

The full version of CML, not of DVC ❓

Reply:
> _If we care about DVC version in the cron job, we'll have to add its full version to the cache key too._

Maybe we can just add `curl --silent https://updater.dvc.org | jq -r .version` to the metadata and cache everything to the exact version?

Reply:
Have d5ff751. Not complex at all :see_no_evil:

Reply:
[Not... complex... at... all...](https://en.wikipedia.org/wiki/Rube_Goldberg_machine) 🙈 🙉 🙊

Reply:
There was logic to everything though. One of this year's most beautiful commits for me, I think. The version parsing of a py package should use py's builtin version parsing. The rest uses neat bach syntax to access and pre-filter a json API. Could use Python for this too if you want... 
4b799ec

Reply:
Not exactly my cup of tea, but it's definitely an ingenious solution. I would dare to say that it's even optimal if we take into account the workflow constraints. 💯